### PR TITLE
Feat: `hosts` option

### PR DESCRIPTION
**Pull request by mrazauskas**
As discussed, here is solution with an array of hosts. What you think?

NEXT: I will work on `getPort(port: number)`. Would be more useful to have `port: number | string`.

### CODE REVIEW

**Comment by mrazauskas** [comment]
Can this be race condition? Can’t get if GitHub is running test concurrently or not.

The same ports are check by test in different files. Locally I managed to get similar failing result. `jest --runInBand` makes everything work smooth. How can we try this out somehow?

**Comment by mrazauskas** [comment]
@pi0 Please take a look.

Tests are fixed. I was rather naive to expect to have `localhost` on Ubuntu.

File: src/index.ts [hunk 1]
```diff
@@ -51,29 +57,45 @@ export async function getPort (config?: GetPortInput): Promise<number> {
...
-    const r = await checkPort(port, host)
-    if (r) {
-      return r
+    for (const host of hosts) {
+      const _port = availablePort[0] || port
+
+      try {
+        availablePort.push(await checkPort(_port, host))
+      } catch (e) {
+        if (e.code !== 'EADDRINUSE') { // this was expected
```
Comment (pi0) [inline src/index.ts:78]: We can move it to `checkPort` `return Promise().catch(err => err.code = EADDRINUSE ? return true : throw err`

File: src/index.ts [hunk 2]
```diff
@@ -51,29 +57,45 @@ export async function getPort (config?: GetPortInput): Promise<number> {
...
 
-async function checkPorts (ports: number[], host: string): Promise<number> {
+async function resolvePort (ports: number[], hosts: string[]): Promise<number> {
+  let availablePort: number[] = []
+
   for (const port of ports) {
-    const r = await checkPort(port, host)
-    if (r) {
-      return r
+    for (const host of hosts) {
```
Comment (pi0) [inline src/index.ts:72]: Perf: we might use `Promise.all` on host check

File: src/index.ts [hunk 3]
```diff
@@ -51,29 +57,45 @@ export async function getPort (config?: GetPortInput): Promise<number> {
...
 }
 
-function checkPort (port: number, host: string): Promise<number|false> {
-  return new Promise((resolve) => {
+function checkPort (port: number, host: string): Promise<number> {
+  return new Promise((resolve, reject) => {
     const server = createServer()
     server.unref()
-    server.on('error', () => { resolve(false) })
+    server.on('error', (e) => { reject(e) })
```
Comment (pi0) [inline src/index.ts:98]: Codestyle: please use `err` for naming

**Review by pi0 — COMMENTED** [review]
Nice changes 👍

**Comment by mrazauskas** [comment]
Thanks. `Promise.all` does great job in this case. Already implemented it.

And noticed that `checkPort(0, host)` returns different value for each host. This does not make sure that random port will not be in use on all the hosts. Working on this.

**Comment by mrazauskas** [comment]
Seems like splitting logic into smaller parts helped to solve everything.

`listen(0, host)` was involving additional logic (take on host, remove it from array, get port, check the others). Too complicated. Seems like is enough to get random number `Math.random()`.

**Comment by mrazauskas** [comment]
Seems like giving up `listen(0, host)` help to solve everything. With several hosts `listen(0, host)` required too much additional logic (take one host, remove it from array, get port, check other hosts). Too complicated.

With `Math.random()` in all the cases we know port number in advance. No need to return it from `listen()`. All we need if `resolve` and `reject` to make `Promise.all` work. It waits until first promise to be rejected and returns `false`. This is indeed performant. Later the boolean values tick a counter inside `do ... while`. Here we have at least one execution (if `ports` array is empty, a random port is check) that’s why `do ... while` is employed.

What you think @pi0?

**Comment by mrazauskas** [comment]
`Promise.all` was great idea. It works smoothly on MacOS, but unfortunately on Ubuntu hosts have to be checked sequentially. For example, on Ubuntu if there is a listener on `0.0.0.0:<port>`, an attempt to listen on `127.0.0.1:<port>` throws `EADDRINUSE: address already in use`. That’s why `Promise.all` was creating infinite loop and in the end Node was crashing with `out of memory`. But I think `Promise.all` helped to shape the code better.

Also giving up `listen(0, host)` helped. If several hosts have to be checked, `listen(0, host)` required too much additional logic (take one host, remove it from array, get port, check other hosts). Too complicated. With `Math.random()` in all the cases we know port number in advance. No need to return it from `listen()`.

What you think @pi0?

**Comment by mrazauskas** [comment]
One more though. After working on it I am not 100% sure that additional `hosts` option is really useful. It makes API surface larger, but is there any benefit? It might be enough to have: `host?: string | string[]` with default value: `process.env.HOST || ['0.0.0.0', '127.0.01']`.

Where `['0.0.0.0', '127.0.01']` is useful? This is very good default on MacOS, but elsewhere a `string` will do the job. Right? If the array is mostly useful as default value, why don’t we simply add it as default value? Or I missed something important?

Shall I refactor?

**Pull request closed without merging**